leksah: add the appropriate deps (don't merge)

obadz opened the pull request:
wrapGAppsHook seems to be called twice for a mysterious reason. I've added a little guard making sure that if that happens, it doesn't just delete the binary, which is what it was doing before. Instead, it will wrap the wrapper.

leksah needs deps on defaultIcontTheme, wrapGAppsHook, gtk3 and a the right path to leksah-server which this commit provides.

Commit:
leksah: add the appropriate deps to enable running

obadz commented:
cc // @lethalman @peti @jaccokrijnen @ocharles

lethalman commented:
That code must not be in `wrapGAppsHook`, and we have to understand why the hook is being called twice despite `$outputs` is just one.

peti commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -635,7 +635,19 @@ self: super: {
+    postPatch = (drv.postPatch or "") + ''
+      for f in src/IDE/Leksah.hs src/IDE/Utils/ServerConnection.hs
+      do
+        substituteInPlace "$f" --replace "\"leksah-server\"" "\"${super.leksah-server}/bin/leksah-server\""
```
You should refer to `self.leksah-server` here. The "super" version lacks the overrides potentially applied in this file.

obadz replied:
Fixed

obadz commented:
Agree that we should identify why it's called twice.

Should I move these lines up to wrapProgram? They must exist somewhere in the code base already as pithos is wrapped twice, once by wrapGAppsHook and once by the python stuff and that somehow just works:

```
$ ls -la /nix/store/*-pithos-*/bin
-r-xr-xr-x 1 root nixbld 1.8K Jan  1  1970 pithos*
-r-xr-xr-x 1 root nixbld 2.8K Jan  1  1970 .pithos-wrapped*
-r-xr-xr-x 1 root nixbld  389 Jan  1  1970 ..pithos-wrapped-wrapped*
```

peti commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -635,7 +635,19 @@ self: super: {
   # Tries to run GUI in tests
-  leksah = dontCheck super.leksah;
+  leksah = dontCheck (overrideCabal super.leksah (drv: {
+    executableHaskellDepends = with pkgs; [
```
`xxxHaskellDepends` is for *Haskell* packages. Listing `gnome3.defaultIconTheme` is almost certainly wrong. I'd suggest adding those entries to the `executableSystemDepends` attribute instead. If they are specific to the test suite (i.e. they aren't required when building with `doCheck=false`), then they should go to `testSystemDepends`.
`wrapGAppsHook` is not actually a library, so this would probably have to be listed in `testToolDepends`.

obadz commented:
@peti's suggestion to move `wrapGAppsHook` to `executableSystemDepends` actually shook things up and might give us a clue as to why it's being run twice:

It used to be run twice at the very end of the build, but now:

```
[..]
hicolorPreFixupPhase
glibPreFixupPhase
post-installation fixup
Wrapping app /nix/store/i3j7iky4h96gwxjm1w2zplsfm2550fkz-leksah-0.15.1.4/bin/bewleksah
Wrapping app /nix/store/i3j7iky4h96gwxjm1w2zplsfm2550fkz-leksah-0.15.1.4/bin/leksah
patching ELF executables and libraries in /nix/store/i3j7iky4h96gwxjm1w2zplsfm2550fkz-leksah-0.15.1.4
/nix/store/i3j7iky4h96gwxjm1w2zplsfm2550fkz-leksah-0.15.1.4/lib/ghc-7.10.2/leksah-0.15.1.4/libHSleksah-0.15.1.4-5cla3Pi2vpI29XeBHQZWta-ghc7.10.2.so
[..]
strip:/nix/store/i3j7iky4h96gwxjm1w2zplsfm2550fkz-leksah-0.15.1.4/bin/bewleksah: File format not recognized
patching script interpreter paths in /nix/store/i3j7iky4h96gwxjm1w2zplsfm2550fkz-leksah-0.15.1.4
Wrapping app /nix/store/i3j7iky4h96gwxjm1w2zplsfm2550fkz-leksah-0.15.1.4/bin/bewleksah
Wrapping app /nix/store/i3j7iky4h96gwxjm1w2zplsfm2550fkz-leksah-0.15.1.4/bin/leksah
END OF BUILD LOG
```

@peti, those are not test deps, so moving to `executableSystemDepends` as suggested.

vcunat commented:
Setup hooks are executed multiple times if they get into build inputs multiple times &ndash; then the hook gets added twice and run twice.

vcunat commented:
There's no deduplication when adding build inputs, as that would be relatively complicated (it's purely bash code). Most hooks are idempotent, so it wouldn't matter.

What about adding a guard variable to each problematic hook?

obadz commented:
@vcunat, one issue is that we're not clear on why this hook is executed twice. I only added the dependency to `wrapGAppsHook` once and it was executed zero times before.

Also, shouldn't `wrapProgram` make sure that `.${file}-wrapped` doesn't exist prior to wrapping `$file`? Seems like in our case `wrapProgram` being invoked twice actually led to the binary being deleted and  replaced by one of the wrappers.

vcunat commented:
I had propagated build inputs in mind, which could lead to adding it multiple times nevertheless, though I'm not sure why would anyone add the hook into *propagated* build inputs. It might be some other cause.

peti commented:
`libraryHaskellDepends` are `propagatedBuildInputs`.

obadz commented:
but now we're using `executableSystemDepends` so should just be `nativeBuildInputs` ?

peti commented:
Yes. `xxxSystemDepends` are not propagated, only `xxxHaskellDepends` are.

obadz commented:
That probably explains why one of the calls to `wrapGAppsHook` "moved" as explained in https://github.com/NixOS/nixpkgs/pull/10351#issuecomment-147414633 but not sure why it's still there twice.

vcunat commented:
I staged a change that should avoid running the hook multiple times.

obadz commented:
@vcunat, that's great, thanks! Could you reference the commit/PR?

vcunat commented:
d10b910 It also was above in the list (if viewing through web).

peti commented:
What happens to this PR now?

obadz commented:
@peti, I'll wait until the change makes its way to the `nixos-unstable`  channel and then revive the PR ?

vcunat commented:
Just note that it might take even weeks to happen with the current state of Hydra overload.

jagajaga commented:
Ping? :)

obadz commented:
I've rebased over the latest `nixos-unstable` and removed the `wrapGAppsHook` hack.

I'm now getting a build error that I do not understand. The log looks like this:

```
strip:/nix/store/1dlzaccxnfnxgxlcm89yhl03qm2zghvi-leksah-0.15.1.4/lib/ghc-7.10.2/leksah-0.15.1.4/IDE/Command.dyn_hi: File format not recognized
strip:/nix/store/1dlzaccxnfnxgxlcm89yhl03qm2zghvi-leksah-0.15.1.4/lib/ghc-7.10.2/leksah-0.15.1.4/IDE/Command.hi: File format not recognized
strip:/nix/store/1dlzaccxnfnxgxlcm89yhl03qm2zghvi-leksah-0.15.1.4/bin/leksah: File format not recognized
strip:/nix/store/1dlzaccxnfnxgxlcm89yhl03qm2zghvi-leksah-0.15.1.4/bin/bewleksah: File format not recognized
patching script interpreter paths in /nix/store/1dlzaccxnfnxgxlcm89yhl03qm2zghvi-leksah-0.15.1.4
builder for ‘/nix/store/wi95akhnjz5qxrq2vmizysdzxw0rjid2-leksah-0.15.1.4.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/4kv5rv3smqimgzbzigmmmsqdzdfwkqcc-ghc-7.10.2.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/4kv5rv3smqimgzbzigmmmsqdzdfwkqcc-ghc-7.10.2.drv’ failed
/run/current-system/sw/bin/nix-shell: failed to build all dependencies
```

Since `strip` failures aren't supposed to be fatal, I'm not sure what the problem is.

The good news: invoking `/nix/store/***-leksah-0.15.1.4/bin/leksah` seems to launch fine. (Oddly it didn't seem to be cleaned out of my `/nix/store` immediately after the failure).

Maybe others want to give it a try..

vcunat commented:
These can be difficult to debug. You can put `set -x` into some of the later phases to make bash vomit lots of information. One can usually find in it what command originally throwed the error silently.

peti commented:
This issue is fixed by 8f83fa04bcb9c06a293d0fa1650fb119ff46191c. The leksah changes are merged in 07a54c9918dbe52544e48afc39815585c234120c. Both will show up on `master` in a moment.

obadz commented:
Nice! Thanks @peti!
